File: client/src/lib/api/monitoring.ts
/**
 * After-hours Monitoring API Client
 * 시간외 거래 모니터링 API 클라이언트
 */

import { apiClient } from './client';
import type {
  MonitoringSessionStatus,
  MonitoringTarget,
  ThresholdAdjustmentRequest,
  ThresholdAdjustmentResponse,
  MarketCondition,
  StartMonitoringRequest
} from '@/types/monitoring';

export class MonitoringAPIError extends Error {
  constructor(message: string, public statusCode?: number) {
    super(message);
    this.name = 'MonitoringAPIError';
  }
}

// Error handling utility
function extractErrorMessage(error: unknown, fallback: string): string {
  if (error && typeof error === 'object' && 'response' in error) {
    const response = (error as { response?: { data?: { detail?: string }; status?: number } }).response;
    if (response?.data?.detail) {
      return response.data.detail;
    }
  }
  return fallback;
}

function extractStatusCode(error: unknown): number | undefined {
  if (error && typeof error === 'object' && 'response' in error) {
    const response = (error as { response?: { status?: number } }).response;
    return response?.status;
  }
  return undefined;
}

export const monitoringAPI = {
  // === Session Management ===

  /**
   * 시간외 모니터링 세션 시작
   */
  async startSession(request: StartMonitoringRequest): Promise<{
    success: boolean;
    message: string;
    targets_count: number;
  }> {
    try {
      const response = await apiClient.post<{
        success: boolean;
        message: string;
        targets_count: number;
      }>('/api/monitoring/start', request);

      if (!response.success || !response.data) {
        throw new MonitoringAPIError(response.error || 'Failed to start monitoring session');
      }

      return response.data;
    } catch (error: unknown) {
      if (error instanceof MonitoringAPIError) {
        throw error;
      }
      throw new MonitoringAPIError(
        extractErrorMessage(error, 'Failed to start monitoring session'),
        extractStatusCode(error)
      );
    }
  },

  /**
   * 시간외 모니터링 세션 중지
   */
  async stopSession(): Promise<{
    success: boolean;
    message: string;
  }> {
    try {
      const response = await apiClient.post<{
        success: boolean;
        message: string;
      }>('/api/monitoring/stop');

      if (!response.success || !response.data) {
        throw new MonitoringAPIError(response.error || 'Failed to stop monitoring session');
      }

      return response.data;
    } catch (error: unknown) {
      if (error instanceof MonitoringAPIError) {
        throw error;
      }
      throw new MonitoringAPIError(
        extractErrorMessage(error, 'Failed to stop monitoring session'),
        extractStatusCode(error)
      );
    }
  },

  /**
   * 모니터링 상태 조회
   */
  async getStatus(): Promise<{
    success: boolean;
    status: MonitoringSessionStatus;
  }> {
    try {
      const response = await apiClient.get<{
        success: boolean;
        status: MonitoringSessionStatus;
      }>('/api/monitoring/status');

      if (!response.success || !response.data) {
        throw new MonitoringAPIError(response.error || 'Failed to get monitoring status');
      }

      return response.data;
    } catch (error: unknown) {
      if (error instanceof MonitoringAPIError) {
        throw error;
      }
      throw new MonitoringAPIError(
        extractErrorMessage(error, 'Failed to get monitoring status'),
        extractStatusCode(error)
      );
    }
  },

  /**
   * 모니터링 대상 목록 조회
   */
  async getTargets(): Promise<{
    success: boolean;
    targets: MonitoringTarget[];
    count: number;
  }> {
    try {
      const response = await apiClient.get<{
        success: boolean;
        targets: MonitoringTarget[];
        count: number;
      }>('/api/monitoring/targets');

      if (!response.success || !response.data) {
        throw new MonitoringAPIError(response.error || 'Failed to get monitoring targets');
      }

      return response.data;
    } catch (error: unknown) {
      if (error instanceof MonitoringAPIError) {
        throw error;
      }
      throw new MonitoringAPIError(
        extractErrorMessage(error, 'Failed to get monitoring targets'),
        extractStatusCode(error)
      );
    }
  },

  // === Threshold Management ===

  /**
   * 특정 종목 임계값 조정
   */
  async adjustThreshold(request: ThresholdAdjustmentRequest): Promise<{
    success: boolean;
    message: string;
    symbol: string;
    new_threshold: number;
    strategy: string;
  }> {
    try {
      const response = await apiClient.post<{
        success: boolean;
        message: string;
        symbol: string;
        new_threshold: number;
        strategy: string;
      }>('/api/monitoring/adjust-threshold', request);

      if (!response.success || !response.data) {
        throw new MonitoringAPIError(response.error || 'Failed to adjust threshold');
      }

      return response.data;
    } catch (error: unknown) {
      if (error instanceof MonitoringAPIError) {
        throw error;
      }
      throw new MonitoringAPIError(
        extractErrorMessage(error, 'Failed to adjust threshold'),
        extractStatusCode(error)
      );
    }
  },

  /**
   * 자동 임계값 조정
   */
  async autoAdjustThresholds(request: {
    strategy: 'conservative' | 'balanced' | 'aggressive' | 'time_based';
    apply_all: boolean;
    target_symbols?: string[];
  }): Promise<{
    success: boolean;
    message: string;
    strategy: string;
    market_condition: MarketCondition;
    adjustments: ThresholdAdjustmentResponse[];
  }> {
    try {
      const response = await apiClient.post<{
        success: boolean;
        message: string;
        strategy: string;
        market_condition: MarketCondition;
        adjustments: ThresholdAdjustmentResponse[];
      }>('/api/monitoring/auto-adjust-thresholds', request);

      if (!response.success || !response.data) {
        throw new MonitoringAPIError(response.error || 'Failed to auto-adjust thresholds');
      }

      return response.data;
    } catch (error: unknown) {
      if (error instanceof MonitoringAPIError) {
        throw error;
      }
      throw new MonitoringAPIError(
        extractErrorMessage(error, 'Failed to auto-adjust thresholds'),
        extractStatusCode(error)
      );
    }
  },

  /**
   * 권장 조정 전략 조회
   */
  async getSuggestedStrategies(): Promise<{
    success: boolean;
    market_condition: MarketCondition;
    suggested_strategies: Array<{
      strategy: string;
      description: string;
    }>;
  }> {
    try {
      const response = await apiClient.get<{
        success: boolean;
        market_condition: MarketCondition;
        suggested_strategies: Array<{
          strategy: string;
          description: string;
        }>;
      }>('/api/monitoring/suggested-strategies');

      if (!response.success || !response.data) {
        throw new MonitoringAPIError(response.error || 'Failed to get suggested strategies');
      }

      return response.data;
    } catch (error: unknown) {
      if (error instanceof MonitoringAPIError) {
        throw error;
      }
      throw new MonitoringAPIError(
        extractErrorMessage(error, 'Failed to get suggested strategies'),
        extractStatusCode(error)
      );
    }
  },

  /**
   * 임계값 조정 미리보기
   */
  async previewThresholdAdjustment(
    strategy: 'conservative' | 'balanced' | 'aggressive' | 'time_based',
    symbol?: string
  ): Promise<{
    success: boolean;
    strategy: string;
    market_condition: MarketCondition;
    previews: Array<{
      symbol: string;
      stock_name: string;
      current_threshold: number;
      recommended_threshold: number;
      adjustment_reason: string;
      confidence_score: number;
    }>;
  }> {
    try {
      const params = new URLSearchParams({ strategy });
      if (symbol) params.append('symbol', symbol);

      const response = await apiClient.get<{
        success: boolean;
        strategy: string;
        market_condition: MarketCondition;
        previews: Array<{
          symbol: string;
          stock_name: string;
          current_threshold: number;
          recommended_threshold: number;
          adjustment_reason: string;
          confidence_score: number;
        }>;
      }>(`/api/monitoring/threshold-preview?${params}`);

      if (!response.success || !response.data) {
        throw new MonitoringAPIError(response.error || 'Failed to preview threshold adjustment');
      }

      return response.data;
    } catch (error: unknown) {
      if (error instanceof MonitoringAPIError) {
        throw error;
      }
      throw new MonitoringAPIError(
        extractErrorMessage(error, 'Failed to preview threshold adjustment'),
        extractStatusCode(error)
      );
    }
  },

  // === Analytics and History ===

  /**
   * 세션 기록 조회
   */
  async getSessionHistory(): Promise<{
    success: boolean;
    current_session: {
      phase: string;
      start_time: string;
      total_targets: number;
      triggered_count: number;
      is_running: boolean;
    };
    history: Array<{
      session_id: string;
      start_time: string;
      end_time?: string;
      total_targets: number;
      triggered_count: number;
      success_rate: number;
    }>; // Session history records
  }> {
    try {
      const response = await apiClient.get<{
        success: boolean;
        current_session: {
          phase: string;
          start_time: string;
          total_targets: number;
          triggered_count: number;
          is_running: boolean;
        };
        history: Array<{
          session_id: string;
          start_time: string;
          end_time?: string;
          total_targets: number;
          triggered_count: number;
          success_rate: number;
        }>;
      }>('/api/monitoring/session-history');

      if (!response.success || !response.data) {
        throw new MonitoringAPIError(response.error || 'Failed to get session history');
      }

      return response.data;
    } catch (error: unknown) {
      if (error instanceof MonitoringAPIError) {
        throw error;
      }
      throw new MonitoringAPIError(
        extractErrorMessage(error, 'Failed to get session history'),
        extractStatusCode(error)
      );
    }
  },

  /**
   * 모니터링 성과 통계
   */
  async getPerformanceStats(): Promise<{
    success: boolean;
    stats: {
      total_targets: number;
      triggered_count: number;
      success_rate: number;
      average_change_percent: number;
      session_duration_minutes: number;
      current_phase: string;
    };
  }> {
    try {
      const response = await apiClient.get<{
        success: boolean;
        stats: {
          total_targets: number;
          triggered_count: number;
          success_rate: number;
          average_change_percent: number;
          session_duration_minutes: number;
          current_phase: string;
        };
      }>('/api/monitoring/performance-stats');

      if (!response.success || !response.data) {
        throw new MonitoringAPIError(response.error || 'Failed to get performance stats');
      }

      return response.data;
    } catch (error: unknown) {
      if (error instanceof MonitoringAPIError) {
        throw error;
      }
      throw new MonitoringAPIError(
        extractErrorMessage(error, 'Failed to get performance stats'),
        extractStatusCode(error)
      );
    }
  },

  // === Real-time Utilities ===

  /**
   * 실시간 모니터링 상태 스트림 (폴링 기반)
   */
  createStatusStream(intervalMs: number = 30000) {
    let intervalId: NodeJS.Timeout | null = null;
    const listeners: Array<(status: MonitoringSessionStatus) => void> = [];

    const start = () => {
      if (intervalId) return;

      const fetchStatus = async () => {
        try {
          const { status } = await this.getStatus();
          listeners.forEach(listener => listener(status));
        } catch (error) {
          console.error('Failed to fetch monitoring status:', error);
        }
      };

      // 즉시 한 번 실행
      fetchStatus();

      // 주기적 실행
      intervalId = setInterval(fetchStatus, intervalMs);
    };

    const stop = () => {
      if (intervalId) {
        clearInterval(intervalId);
        intervalId = null;
      }
    };

    const subscribe = (listener: (status: MonitoringSessionStatus) => void) => {
      listeners.push(listener);
      return () => {
        const index = listeners.indexOf(listener);
        if (index > -1) {
          listeners.splice(index, 1);
        }
      };
    };

    return { start, stop, subscribe };
  },

  /**
   * 임계값 조정 배치 적용
   */
  async batchAdjustThresholds(adjustments: Array<{
    symbol: string;
    new_threshold: number;
  }>): Promise<Array<{
    symbol: string;
    success: boolean;
    error?: string;
  }>> {
    const results = await Promise.allSettled(
      adjustments.map(async ({ symbol, new_threshold }) => {
        try {
          await this.adjustThreshold({
            symbol,
            new_threshold,
            strategy: 'manual'
          });
          return { symbol, success: true };
        } catch (error) {
          return {
            symbol,
            success: false,
            error: error instanceof Error ? error.message : 'Unknown error'
          };
        }
      })
    );

    return results.map((result, index) => {
      const symbol = adjustments[index].symbol;
      if (result.status === 'fulfilled') {
        return result.value;
      } else {
        return {
          symbol,
          success: false,
          error: result.reason?.message || 'Failed to adjust threshold'
        };
      }
    });
  }
};

export default monitoringAPI;

File: server/app/api/endpoints/monitoring.py
"""
After-hours Monitoring API Endpoints
시간외 거래 모니터링 API 엔드포인트
"""

from typing import List, Dict, Any, Optional
from fastapi import APIRouter, HTTPException, Depends
from pydantic import BaseModel, Field
from datetime import datetime

from ...core.monitoring.session_manager import session_manager, SessionPhase, MonitoringTarget, SessionStatus
from ...core.monitoring.threshold_adjuster import threshold_adjuster, AdjustmentStrategy, MarketCondition
from ...services.kis_api import KISAPIClient

router = APIRouter()


# Request Models
class StartMonitoringRequest(BaseModel):
    """모니터링 시작 요청"""
    targets: List[Dict[str, Any]] = Field(description="모니터링 대상 주식 목록")
    auto_start: bool = Field(default=True, description="자동 시작 여부")


class AdjustThresholdRequest(BaseModel):
    """임계값 조정 요청"""
    symbol: str = Field(min_length=1, max_length=10)
    new_threshold: float = Field(gt=0, le=10)
    strategy: str = Field(default="manual", pattern="^(manual|conservative|balanced|aggressive|time_based)$")


class ThresholdAdjustmentRequest(BaseModel):
    """임계값 자동 조정 요청"""
    strategy: str = Field(pattern="^(conservative|balanced|aggressive|time_based)$")
    apply_all: bool = Field(default=False, description="모든 종목에 적용 여부")
    target_symbols: Optional[List[str]] = Field(default=None, description="특정 종목만 적용")


# Response Models
class MonitoringStatusResponse(BaseModel):
    """모니터링 상태 응답"""
    is_running: bool
    current_phase: str
    phase_start_time: str
    next_phase_time: Optional[str]
    monitoring_targets: List[Dict[str, Any]]
    total_targets: int
    triggered_count: int
    remaining_time_seconds: int


class ThresholdAdjustmentResponse(BaseModel):
    """임계값 조정 응답"""
    symbol: str
    old_threshold: float
    new_threshold: float
    adjustment_reason: str
    confidence_score: float
    strategy: str


# === Monitoring Session Management APIs ===

@router.post("/monitoring/start")
async def start_monitoring_session(request: StartMonitoringRequest):
    """시간외 모니터링 세션 시작"""
    try:
        # 입력 검증
        if not request.targets:
            raise HTTPException(status_code=400, detail="No monitoring targets provided")

        # 필수 필드 검증
        for target in request.targets:
            required_fields = ["symbol", "stock_name", "entry_price"]
            for field in required_fields:
                if field not in target:
                    raise HTTPException(status_code=400, detail=f"Missing required field: {field}")

        # 세션 시작
        success = await session_manager.start_session(request.targets)

        if not success:
            raise HTTPException(status_code=400, detail="Failed to start monitoring session")

        return {
            "success": True,
            "message": f"Monitoring session started with {len(request.targets)} targets",
            "targets_count": len(request.targets)
        }

    except Exception as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.post("/monitoring/stop")
async def stop_monitoring_session():
    """시간외 모니터링 세션 중지"""
    try:
        await session_manager.stop_session()

        return {
            "success": True,
            "message": "Monitoring session stopped"
        }

    except Exception as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.get("/monitoring/status")
async def get_monitoring_status():
    """모니터링 상태 조회"""
    try:
        status = await session_manager.get_session_status()

        return {
            "success": True,
            "status": {
                "is_running": session_manager.is_running,
                "current_phase": status.current_phase.value,
                "phase_start_time": status.phase_start_time.isoformat(),
                "next_phase_time": status.next_phase_time.strftime('%H:%M') if status.next_phase_time else None,
                "monitoring_targets": [
                    {
                        "symbol": target.symbol,
                        "stock_name": target.stock_name,
                        "entry_price": target.entry_price,
                        "current_price": target.current_price,
                        "change_percent": target.change_percent,
                        "volume": target.volume,
                        "buy_threshold": target.buy_threshold,
                        "is_triggered": target.is_triggered,
                        "trigger_time": target.trigger_time.isoformat() if target.trigger_time else None
                    } for target in status.monitoring_targets
                ],
                "total_targets": status.total_targets,
                "triggered_count": status.triggered_count,
                "remaining_time_seconds": int(status.remaining_time.total_seconds())
            }
        }

    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/monitoring/targets")
async def get_monitoring_targets():
    """모니터링 대상 목록 조회"""
    try:
        targets = []
        for symbol, target in session_manager.monitoring_targets.items():
            targets.append({
                "symbol": target.symbol,
                "stock_name": target.stock_name,
                "entry_price": target.entry_price,
                "current_price": target.current_price,
                "change_percent": target.change_percent,
                "volume": target.volume,
                "buy_threshold": target.buy_threshold,
                "is_triggered": target.is_triggered,
                "trigger_time": target.trigger_time.isoformat() if target.trigger_time else None
            })

        return {
            "success": True,
            "targets": targets,
            "count": len(targets)
        }

    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


# === Threshold Management APIs ===

@router.post("/monitoring/adjust-threshold")
async def adjust_threshold(request: AdjustThresholdRequest):
    """특정 종목 임계값 조정"""
    try:
        success = await session_manager.adjust_threshold(
            symbol=request.symbol,
            new_threshold=request.new_threshold
        )

        if not success:
            raise HTTPException(status_code=404, detail=f"Target not found: {request.symbol}")

        return {
            "success": True,
            "message": f"Threshold adjusted for {request.symbol}",
            "symbol": request.symbol,
            "new_threshold": request.new_threshold,
            "strategy": request.strategy
        }

    except Exception as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.post("/monitoring/auto-adjust-thresholds")
async def auto_adjust_thresholds(request: ThresholdAdjustmentRequest):
    """자동 임계값 조정"""
    try:
        # 현재 시장 상황 분석
        kis_client = KISAPIClient()
        market_data = await kis_client.get_market_overview()

        if not market_data:
            market_data = []

        market_condition = threshold_adjuster.get_market_condition_analysis(market_data)

        # 조정 대상 종목 결정
        if request.apply_all:
            target_symbols = list(session_manager.monitoring_targets.keys())
        elif request.target_symbols:
            target_symbols = request.target_symbols
        else:
            raise HTTPException(status_code=400, detail="No target symbols specified")

        # 조정 결과 저장
        adjustment_results = []

        for symbol in target_symbols:
            if symbol not in session_manager.monitoring_targets:
                continue

            target = session_manager.monitoring_targets[symbol]
            current_threshold = target.buy_threshold

            # 조정 계산
            strategy = AdjustmentStrategy(request.strategy)
            recommendation = threshold_adjuster.calculate_adjustment(
                current_threshold=current_threshold,
                market_condition=market_condition,
                current_time=datetime.now(),
                strategy=strategy
            )

            # 임계값 적용
            await session_manager.adjust_threshold(symbol, recommendation.recommended_threshold)

            adjustment_results.append({
                "symbol": symbol,
                "old_threshold": current_threshold,
                "new_threshold": recommendation.recommended_threshold,
                "adjustment_reason": recommendation.adjustment_reason,
                "confidence_score": recommendation.confidence_score,
                "strategy": strategy.value
            })

        return {
            "success": True,
            "message": f"Auto-adjusted thresholds for {len(adjustment_results)} targets",
            "strategy": request.strategy,
            "market_condition": {
                "total_rise_count": market_condition.total_rise_count,
                "total_stock_count": market_condition.total_stock_count,
                "average_change": market_condition.average_change,
                "volatility_index": market_condition.volatility_index,
                "volume_ratio": market_condition.volume_ratio
            },
            "adjustments": adjustment_results
        }

    except Exception as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.get("/monitoring/suggested-strategies")
async def get_suggested_strategies():
    """권장 조정 전략 조회"""
    try:
        # 현재 시장 상황 분석
        kis_client = KISAPIClient()
        market_data = await kis_client.get_market_overview()

        if not market_data:
            market_data = []

        market_condition = threshold_adjuster.get_market_condition_analysis(market_data)
        strategies = threshold_adjuster.get_suggested_strategies(market_condition)

        return {
            "success": True,
            "market_condition": {
                "total_rise_count": market_condition.total_rise_count,
                "total_stock_count": market_condition.total_stock_count,
                "average_change": market_condition.average_change,
                "volatility_index": market_condition.volatility_index,
                "volume_ratio": market_condition.volume_ratio
            },
            "suggested_strategies": [
                {
                    "strategy": strategy.value,
                    "description": description
                } for strategy, description in strategies
            ]
        }

    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/monitoring/threshold-preview")
async def preview_threshold_adjustment(
    strategy: str,
    symbol: Optional[str] = None
):
    """임계값 조정 미리보기"""
    try:
        if strategy not in ["conservative", "balanced", "aggressive", "time_based"]:
            raise HTTPException(status_code=400, detail="Invalid strategy")

        # 현재 시장 상황 분석
        kis_client = KISAPIClient()
        market_data = await kis_client.get_market_overview()

        if not market_data:
            market_data = []

        market_condition = threshold_adjuster.get_market_condition_analysis(market_data)

        # 미리보기 계산
        previews = []

        if symbol:
            # 특정 종목만
            if symbol in session_manager.monitoring_targets:
                target = session_manager.monitoring_targets[symbol]
                recommendation = threshold_adjuster.calculate_adjustment(
                    current_threshold=target.buy_threshold,
                    market_condition=market_condition,
                    current_time=datetime.now(),
                    strategy=AdjustmentStrategy(strategy)
                )
                previews.append({
                    "symbol": symbol,
                    "stock_name": target.stock_name,
                    "current_threshold": target.buy_threshold,
                    "recommended_threshold": recommendation.recommended_threshold,
                    "adjustment_reason": recommendation.adjustment_reason,
                    "confidence_score": recommendation.confidence_score
                })
        else:
            # 모든 종목
            for symbol, target in session_manager.monitoring_targets.items():
                recommendation = threshold_adjuster.calculate_adjustment(
                    current_threshold=target.buy_threshold,
                    market_condition=market_condition,
                    current_time=datetime.now(),
                    strategy=AdjustmentStrategy(strategy)
                )
                previews.append({
                    "symbol": symbol,
                    "stock_name": target.stock_name,
                    "current_threshold": target.buy_threshold,
                    "recommended_threshold": recommendation.recommended_threshold,
                    "adjustment_reason": recommendation.adjustment_reason,
                    "confidence_score": recommendation.confidence_score
                })

        return {
            "success": True,
            "strategy": strategy,
            "market_condition": {
                "total_rise_count": market_condition.total_rise_count,
                "total_stock_count": market_condition.total_stock_count,
                "average_change": market_condition.average_change,
                "volatility_index": market_condition.volatility_index
            },
            "previews": previews
        }

    except Exception as e:
        raise HTTPException(status_code=400, detail=str(e))


# === Session History and Analytics ===

@router.get("/monitoring/session-history")
async def get_session_history():
    """세션 기록 조회"""
    try:
        # 현재는 단순한 현재 세션 정보만 반환
        # 향후 데이터베이스나 파일 기반 기록 시스템 구현 가능

        current_status = await session_manager.get_session_status()

        return {
            "success": True,
            "current_session": {
                "phase": current_status.current_phase.value,
                "start_time": current_status.phase_start_time.isoformat(),
                "total_targets": current_status.total_targets,
                "triggered_count": current_status.triggered_count,
                "is_running": session_manager.is_running
            },
            "history": []  # 향후 구현
        }

    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/monitoring/performance-stats")
async def get_monitoring_performance():
    """모니터링 성과 통계"""
    try:
        targets = list(session_manager.monitoring_targets.values())
        total_targets = len(targets)
        triggered_count = sum(1 for target in targets if target.is_triggered)

        if total_targets == 0:
            success_rate = 0.0
            avg_change = 0.0
        else:
            success_rate = (triggered_count / total_targets) * 100
            avg_change = sum(target.change_percent for target in targets) / total_targets

        return {
            "success": True,
            "stats": {
                "total_targets": total_targets,
                "triggered_count": triggered_count,
                "success_rate": round(success_rate, 2),
                "average_change_percent": round(avg_change, 2),
                "session_duration_minutes": int((datetime.now() - session_manager._get_phase_start_time()).total_seconds() / 60),
                "current_phase": session_manager.current_phase.value
            }
        }

    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))